legacy/builder/container_find_includes.go

```diff
@@ -295,6 +334,46 @@ func writeCache(cache *includeCache, path *paths.Path) error {
 	return nil
 }
 
+// Preload the cached include files and libraries
+func preloadCachedFolder(ctx *types.Context, cache *includeCache) {
+	var entryToRemove []string
+	cache.entries.Range(func(include, entry interface{}) bool {
+
+		header, ok := include.(string)
+		if(ok) {
+			// Ignore the pre-added folder
+			if(!strings.HasPrefix(header, "no_resolve")) {
+				library, imported := ResolveLibrary(ctx, header)
+				if library == nil {
+					if !imported {
+						// Cannot find the library and it is not imported, is it gone? Remove it later
+						entryToRemove = append(entryToRemove, header)
+						cache.valid = false
+					}
+				} else {
+
+					// Add this library to the list of libraries, the
+					// include path and queue its source files for further
+					// include scanning
+					ctx.ImportedLibraries = append(ctx.ImportedLibraries, library)
+					// Since it is already in cache, append the include folder only
+					appendIncludeFolderWoCache(ctx, header, library.SourceDir)
+					sourceDirs := library.SourceDirs()
+					for _, sourceDir := range sourceDirs {
+						queueSourceFilesFromFolder(ctx, ctx.CollectedSourceFiles, library, sourceDir.Dir, sourceDir.Recurse)
+					}
+				}
+			}
+		}
+		return true
+	})
```

Comment:
If I read this correctly, you're basically reusing all the previous set of include paths, but without keeping the history of how you obtained that set. This won't work in all cases, in particular, if the cache is invalidated you must redo all the work again to be sure to not pick libraries that are not needed anymore.

Reply:
Yes, in the particular case you mentioned, if need to clear the useless libraries, a clean build is required to do the job.